# Test task

In [1]:
# import related libraries

import pandas as pd
import numpy as np

#-----------------------------------
import warnings
warnings.filterwarnings("ignore")

In [2]:
# upload data from a relative path
data = pd.read_csv(r"int20h-ds-test-dataset.csv")
data.head() 

userid user_state                  event_name  \
0  c95c777785faec8dd910d019d7278ebe         CA         Add Vehicle Success   
1  c95c777785faec8dd910d019d7278ebe         CA           Add Vehicle Break   
2  f344be2d9a042b7444f3cc5279e38ef1         FL             Calculator View   
3  c95c777785faec8dd910d019d7278ebe         CA  Add Payment Method Success   
4  e331ed81422d8fba55520a43a872e701         IL             Sign Up Success   

                                    event_attributes   event_created_date  \
0  {"Make":"Dodge","Model":"Caravan","Color":"Whi...  2022-01-16 17:03:04   
1                                                 {}  2022-01-16 17:07:47   
2                                                 {}  2022-01-16 17:16:25   
3     {"Payment Method":"Credit","Tokenized Pay":""}  2022-01-16 17:24:22   
4                                 {"Method":"Apple"}  2022-01-16 17:34:51   

  event_platform device_manufacture device_model  
0        android            samsung     SM-N975U  
1        android            samsung     SM-N975U  
2        android            samsung    SM-G973U1  
3        android            samsung     SM-N975U  
4            ios              Apple   iPhone12,1

## Data preprocessing

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23357 entries, 0 to 23356
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   userid              23357 non-null  object
 1   user_state          23286 non-null  object
 2   event_name          23357 non-null  object
 3   event_attributes    23357 non-null  object
 4   event_created_date  23357 non-null  object
 5   event_platform      23357 non-null  object
 6   device_manufacture  15847 non-null  object
 7   device_model        15847 non-null  object
dtypes: object(8)
memory usage: 1.4+ MB


### Dropping duplicates

As we can see event_created_date has Dtype object and using reference to pandas docs [drop duplicates method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html), we can just delete repeating rows

In [4]:
data.drop_duplicates(inplace=True)

### Filling NA values

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22039 entries, 0 to 23356
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   userid              22039 non-null  object
 1   user_state          21968 non-null  object
 2   event_name          22039 non-null  object
 3   event_attributes    22039 non-null  object
 4   event_created_date  22039 non-null  object
 5   event_platform      22039 non-null  object
 6   device_manufacture  15809 non-null  object
 7   device_model        15809 non-null  object
dtypes: object(8)
memory usage: 1.5+ MB


We can see that we have missing values into columns user_state, device_manufacture and device_model. By the way, the number of missing values in device_manufacture, device_model is the same. So we can check a theory about missing these values in some specific cases

In [6]:
column_name = 'device_manufacture'
data[data[column_name].isna()]

userid user_state  \
23     c664cc762bdcb88230b4b4f07e3b7d2e         CA   
34     017f4557810b3ba165d2959c17b916b6         TX   
72     0d51394c39dbf15021ccd09155ee6257         FL   
74     0d51394c39dbf15021ccd09155ee6257         FL   
85     c67290d72ec2974aba0f3975c1773405         FL   
...                                 ...        ...   
23337  189adece0dd21eb905f1bfcc09298ecd         CA   
23338  81d7b7d89e06767d80a1aba25e4e3609         CA   
23345  2c597755067d64bf12050cbd7b62c30f         MD   
23347  d545e0069a53fcc868e01726663cd449         CA   
23356  679eba26c4e75e0afb178360becfa21b         CA   

                        event_name  \
23              Transaction Refund   
34            Subscription Premium   
72                           Order   
74                           Order   
85            Subscription Premium   
...                            ...   
23337                        Order   
23338                        Order   
23345  Subscription Premium Cancel   
23347  Subscription Premium Cancel   
23356         Subscription Premium   

                                        event_attributes   event_created_date  \
23     {"Id":"708ee8f2012042b121be7dcdd6df256a","Amou...  2022-01-16 21:41:05   
34     {"Start Date":"2022-01-16","Renewal Date":"202...  2022-01-17 00:04:58   
72                         {"Premium Membership":"True"}  2022-01-15 09:23:48   
74                         {"Premium Membership":"True"}  2022-01-15 10:16:12   
85     {"Start Date":"2022-01-16","Renewal Date":"202...  2022-01-17 01:31:29   
...                                                  ...                  ...   
23337                      {"Premium Membership":"True"}  2022-04-16 10:56:36   
23338                      {"Premium Membership":"True"}  2022-04-16 10:57:42   
23345                                                 {}  2022-04-16 14:15:24   
23347                                                 {}  2022-04-16 14:50:13   
23356  {"Start Date":"2022-04-16","Renewal Date":"202...  2022-04-16 20:50:39   

      event_platform device_manufacture device_model  
23         outofband                NaN          NaN  
34         outofband                NaN          NaN  
72         outofband                NaN          NaN  
74         outofband                NaN          NaN  
85         outofband                NaN          NaN  
...              ...                ...          ...  
23337      outofband                NaN          NaN  
23338      outofband                NaN          NaN  
23345      outofband                NaN          NaN  
23347      outofband                NaN          NaN  
23356      outofband                NaN          NaN  

[6230 rows x 8 columns]

As we see, we have 6230 cases, and they has one common thing: event_platform is outofband. Let's check the situation where this condition is true

In [7]:
data.query("event_platform == 'outofband'")

userid user_state  \
23     c664cc762bdcb88230b4b4f07e3b7d2e         CA   
34     017f4557810b3ba165d2959c17b916b6         TX   
72     0d51394c39dbf15021ccd09155ee6257         FL   
74     0d51394c39dbf15021ccd09155ee6257         FL   
85     c67290d72ec2974aba0f3975c1773405         FL   
...                                 ...        ...   
23337  189adece0dd21eb905f1bfcc09298ecd         CA   
23338  81d7b7d89e06767d80a1aba25e4e3609         CA   
23345  2c597755067d64bf12050cbd7b62c30f         MD   
23347  d545e0069a53fcc868e01726663cd449         CA   
23356  679eba26c4e75e0afb178360becfa21b         CA   

                        event_name  \
23              Transaction Refund   
34            Subscription Premium   
72                           Order   
74                           Order   
85            Subscription Premium   
...                            ...   
23337                        Order   
23338                        Order   
23345  Subscription Premium Cancel   
23347  Subscription Premium Cancel   
23356         Subscription Premium   

                                        event_attributes   event_created_date  \
23     {"Id":"708ee8f2012042b121be7dcdd6df256a","Amou...  2022-01-16 21:41:05   
34     {"Start Date":"2022-01-16","Renewal Date":"202...  2022-01-17 00:04:58   
72                         {"Premium Membership":"True"}  2022-01-15 09:23:48   
74                         {"Premium Membership":"True"}  2022-01-15 10:16:12   
85     {"Start Date":"2022-01-16","Renewal Date":"202...  2022-01-17 01:31:29   
...                                                  ...                  ...   
23337                      {"Premium Membership":"True"}  2022-04-16 10:56:36   
23338                      {"Premium Membership":"True"}  2022-04-16 10:57:42   
23345                                                 {}  2022-04-16 14:15:24   
23347                                                 {}  2022-04-16 14:50:13   
23356  {"Start Date":"2022-04-16","Renewal Date":"202...  2022-04-16 20:50:39   

      event_platform device_manufacture device_model  
23         outofband                NaN          NaN  
34         outofband                NaN          NaN  
72         outofband                NaN          NaN  
74         outofband                NaN          NaN  
85         outofband                NaN          NaN  
...              ...                ...          ...  
23337      outofband                NaN          NaN  
23338      outofband                NaN          NaN  
23345      outofband                NaN          NaN  
23347      outofband                NaN          NaN  
23356      outofband                NaN          NaN  

[6230 rows x 8 columns]

Here we also have 6230 rows with the same structure as in dataset upper it, so let's fill this na values with outofband, so we can understand the origin of values in that columns

In [8]:
data.loc[data.query("event_platform == 'outofband'").index] = \
data.loc[data.query("event_platform == 'outofband'").index].fillna('outofband')

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22039 entries, 0 to 23356
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   userid              22039 non-null  object
 1   user_state          21968 non-null  object
 2   event_name          22039 non-null  object
 3   event_attributes    22039 non-null  object
 4   event_created_date  22039 non-null  object
 5   event_platform      22039 non-null  object
 6   device_manufacture  22039 non-null  object
 7   device_model        22039 non-null  object
dtypes: object(8)
memory usage: 2.0+ MB


We can see that we didn't trigger any missing value user_state and filled values only in device_manufacture and device_model

Now we can check rows with missing user_state value

In [10]:
column_name = 'user_state' 
data[data[column_name].isna()][:50]

userid user_state  \
20     c5cf3fc44b606d4614a834a5306e9616        NaN   
93     ed09ebeba14600a9ff4defc64e32503c        NaN   
121    9182e94800ca0f8e9094ea3b8f5b169c        NaN   
148    838db409e7565bfa896beff8bf7fe023        NaN   
150    838db409e7565bfa896beff8bf7fe023        NaN   
151    838db409e7565bfa896beff8bf7fe023        NaN   
975    54621802fbf69592d1972e7fb7ad19ca        NaN   
3308   2971d0f3a5edd728ac036117268e4bf5        NaN   
4522   7b11702f547fe296f986cfd01c328158        NaN   
4699   7e1beebdff05874880acc85d7b2389da        NaN   
4985   20862cf1d9170e71956a06b5148625f9        NaN   
5344   e2ada799ff596c27f01eb4fbffe96722        NaN   
5356   7b11702f547fe296f986cfd01c328158        NaN   
6386   3de3f2384c698f5c457c0c0e8814111b        NaN   
6454   1f6449b72b7011ed50616ab35653ce5c        NaN   
6455   c35fdde0e72ec2e2aa24fdc641b5f393        NaN   
6855   e9a73c04eb3a65f55655cf85f068aadc        NaN   
6856   e96df62d3bc821b7faeb38b94806c2cf        NaN   
6857   e96df62d3bc821b7faeb38b94806c2cf        NaN   
7200   b3ea7ae9ac15c9e7946abd485c410139        NaN   
7275   b3ea7ae9ac15c9e7946abd485c410139        NaN   
7276   b3ea7ae9ac15c9e7946abd485c410139        NaN   
7650   651861fb7278cbc318e4535c0bc5d85f        NaN   
7937   7e0ff0bc3bc585d42258c083f7abbd3b        NaN   
7940   16496fa03086eef8353659a7b7f62d6d        NaN   
10583  7b77fab4a3c1e880e0d4a5f19717e260        NaN   
10739  b16e9e7932febc892e045738e8e2fb82        NaN   
10845  6fa324ff4238baf4ddc4a0328c6dd0f0        NaN   
11760  d9f6078c032ef4dcbed653d4187ef17e        NaN   
12127  20862cf1d9170e71956a06b5148625f9        NaN   
13958  9bbc64b89418f683ce7b2fc964185d6b        NaN   
14615  683877e342a942bb4c63c5a95f4a0b7d        NaN   
15298  40968dcccbf32d4aafff3f8e52375526        NaN   
15299  8036b265b00fc913ab7c1d99f97fa15c        NaN   
15420  339d4fd39a60e6d5ebd754a755d3cf5f        NaN   
16172  c380cd1c1953776c15c4cf8f11ab8226        NaN   
16472  5917457bb919cf2a2dffe95d62116543        NaN   
16540  23ee00bc49d179887625f111fb0e0b22        NaN   
16701  1504d445b59141feceee53e240a83058        NaN   
16821  c41b93b5910899817515d100c16fd8f5        NaN   
17490  53d3de5bc7bb2d682f25448284fe7c5a        NaN   
17936  6b4ba796f73c41d841a74a5da47ff6d4        NaN   
18093  2a68f0ca02b554fda40bee45a232209a        NaN   
18094  6ff1dc37039b9bbdd1741e63d22da7e5        NaN   
18340  709be88461646b12d751db1cfdcdc07f        NaN   
18538  3ebda279cc24a52546cfa407c8fea22a        NaN   
18748  46c26d6a40c89571854b061e310f69f3        NaN   
18861  e7d938550e860274c3faecb7e857ed09        NaN   
19210  4d8f8c7836398fe04462177387f2df3c        NaN   
19211  4d8f8c7836398fe04462177387f2df3c        NaN   

                       event_name  \
20                Sign Up Success   
93                Sign Up Success   
121               Sign Up Success   
148               Sign Up Success   
150    Email Confirmation Success   
151           Add Vehicle Success   
975               Sign Up Success   
3308              Sign Up Success   
4522              Sign Up Success   
4699              Sign Up Success   
4985              Sign Up Success   
5344              Sign Up Success   
5356                     Sign Out   
6386              Sign Up Success   
6454              Sign Up Success   
6455              Sign Up Success   
6855              Sign Up Success   
6856              Sign Up Success   
6857                     Sign Out   
7200                Sign Up Error   
7275                Sign Up Error   
7276              Sign Up Success   
7650              Sign Up Success   
7937              Sign Up Success   
7940              Sign Up Success   
10583             Sign Up Success   
10739             Sign Up Success   
10845  Add Payment Method Success   
11760             Sign Up Success   
12127                    Sign Out   
13958             Sign Up Success   
14615             Sign Up Success   
15298             Sign Up Success   
15299             Si

We don't see a specific behaviour here, so let's do it in other way

In [11]:
# count amount of unique user_state values

data.user_state.value_counts()

CA    8371
TX    5997
FL    1873
IL    1120
NY     931
NJ     824
VA     636
PA     234
MD     230
AZ     229
MA     217
OK     117
IN     103
NC      97
WI      97
DE      78
MI      71
MO      64
CT      56
AR      55
CO      51
OH      43
NH      42
NV      42
WV      31
GA      28
IA      27
SC      27
NM      26
DC      26
UT      26
AL      26
TN      25
KY      23
MN      23
OR      21
ME      13
KS      13
VT      13
MT       9
RI       7
ID       5
NE       5
WA       4
HI       3
SD       3
AK       2
LA       2
MS       2
Name: user_state, dtype: int64

We will leave only 20 states with the most amount of operations, for left values we create a special class 'other'

In [12]:
top_states = data.user_state.value_counts()[:20].keys().to_list()
data.user_state = data.user_state.apply(lambda x: x if x in top_states else 'other')

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22039 entries, 0 to 23356
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   userid              22039 non-null  object
 1   user_state          22039 non-null  object
 2   event_name          22039 non-null  object
 3   event_attributes    22039 non-null  object
 4   event_created_date  22039 non-null  object
 5   event_platform      22039 non-null  object
 6   device_manufacture  22039 non-null  object
 7   device_model        22039 non-null  object
dtypes: object(8)
memory usage: 2.0+ MB


Now we have no missing values

#### Processing sample outliers

In [14]:
data.device_model.value_counts()[:50]

outofband             6230
iPhone12,1            1489
iPhone13,4            1386
iPhone14,3             909
iPhone11,8             671
iPhone13,2             645
iPhone12,5             615
iPhone13,3             588
iPhone14,5             481
iPhone14,2             443
iPhone12,8             419
iPhone12,3             361
SM-G998U               281
iPhone13,1             261
SM-G991U               259
SM-G970U               243
iPhone11,6             231
SM-G781U               211
SM-A326U               208
SM-G975U               163
iPhone10,2             154
SM-A205U               153
SM-A115U               148
SM-A716U               141
SM-N986U               141
SM-G965U               135
iPhone11,2             132
moto g play (2021)     117
iPhone10,5             116
SM-N960U               111
SM-F926U               109
iPhone10,1             100
SM-G960U                99
SM-N975U1               98
SM-G996U                97
SM-A125U                94
Pixel 5                 92
S

We can see that we have a lot of 'versions' of Iphone 11, 12 etc, so we can just combine that values in one unique. Also we will do this process to Motorola phones, iPads, Pixel and [Samsung](https://en.wikipedia.org/wiki/Samsung_Galaxy#Release_history)

In [15]:
for iphone_number in range(7, 15):
    data.device_model.loc[data['device_model'].apply(lambda x: f"iphone{iphone_number}" in str(x).lower())] = \
    'iPhone' + str(iphone_number)
data.device_model.loc[data['device_model'].apply(lambda x: (str(x).lower()).startswith('sm'))] = data['device_model'].str[:5]
data.device_model.loc[data['device_model'].apply(lambda x: (str(x).lower()).startswith('pixel'))] = data['device_model'].str[:7]
data.device_model.loc[data['device_model'].apply(lambda x: "ipad" in str(x).lower())] = 'iPad'
data.device_model.loc[data['device_model'].apply(lambda x: "moto" in str(x).lower())] = 'moto'

Now let's view changed list

In [16]:
data.device_model.value_counts()[:50]

outofband          6230
iPhone12           2884
iPhone13           2880
iPhone14           1900
SM-G9              1897
iPhone11           1043
SM-N9               745
iPhone10            525
moto                505
SM-G7               333
SM-A1               313
SM-A3               233
iPhone9             210
SM-A2               173
SM-A5               151
SM-A7               148
SM-S9               124
SM-A0               123
iPhone8             119
SM-F9               118
Pixel 5             113
Nokia X100           89
Pixel 6              83
Pixel 3              82
REVVL V+ 5G          81
LM-Q730              80
Pixel 4              74
DE2118               61
SM-F7                60
LM-V600              60
LM-K300              57
ONEPLUS A6013        32
LM-V350              29
SM-J7                27
iPad                 25
LM-K500              24
SM-G8                20
LM-V405              17
BE2025               15
TMRVL4G              14
Nokia C2 Tennen      13
SM-S2           

We can mark values with amount < 25 by the name of device_manufacture

In [17]:
top_model = data.device_model.value_counts()[:35].keys().to_list()
data.device_model = data.device_model. \
apply(lambda x: x if x in top_model else data.loc[data.device_model == x].iloc[0].device_manufacture)

Let's see all processed data now

In [18]:
data.describe()

userid user_state event_name  \
count                              22039      22039      22039   
unique                              3540         21         23   
top     627f50253b42607513a1c93bb68201ad         CA      Order   
freq                                 320       8371       4845   

       event_attributes   event_created_date event_platform  \
count             22039                22039          22039   
unique             2309                20452              3   
top                  {}  2022-04-10 06:26:47            ios   
freq               4668                    6           9586   

       device_manufacture device_model  
count               22039        22039  
unique                 23           55  
top                 Apple    outofband  
freq                 9586         6230

#### Time series work

In [19]:
# converting event_created_date to pd.datetime object
data.event_created_date = pd.to_datetime(data.event_created_date)

In [20]:
# adding column month with month of operation
data["month"] = data.event_created_date.dt.month

In [21]:
data

userid user_state  \
0      c95c777785faec8dd910d019d7278ebe         CA   
1      c95c777785faec8dd910d019d7278ebe         CA   
2      f344be2d9a042b7444f3cc5279e38ef1         FL   
3      c95c777785faec8dd910d019d7278ebe         CA   
4      e331ed81422d8fba55520a43a872e701         IL   
...                                 ...        ...   
23352  679eba26c4e75e0afb178360becfa21b         CA   
23353  679eba26c4e75e0afb178360becfa21b         CA   
23354  679eba26c4e75e0afb178360becfa21b         CA   
23355  679eba26c4e75e0afb178360becfa21b         CA   
23356  679eba26c4e75e0afb178360becfa21b         CA   

                       event_name  \
0             Add Vehicle Success   
1               Add Vehicle Break   
2                 Calculator View   
3      Add Payment Method Success   
4                 Sign Up Success   
...                           ...   
23352  Add Payment Method Success   
23353  Account Setup Profile Skip   
23354  Account Setup Profile Skip   
23355    Chat Conversation Opened   
23356        Subscription Premium   

                                        event_attributes  event_created_date  \
0      {"Make":"Dodge","Model":"Caravan","Color":"Whi... 2022-01-16 17:03:04   
1                                                     {} 2022-01-16 17:07:47   
2                                                     {} 2022-01-16 17:16:25   
3         {"Payment Method":"Credit","Tokenized Pay":""} 2022-01-16 17:24:22   
4                                     {"Method":"Apple"} 2022-01-16 17:34:51   
...                                                  ...                 ...   
23352  {"Payment Method":"Credit","Tokenized Pay":"",... 2022-04-16 20:49:24   
23353                               {"Screen":"Address"} 2022-04-16 20:50:05   
23354                          {"Screen":"Phone Number"} 2022-04-16 20:50:10   
23355         {"From":"Dashboard","Transaction type":""} 2022-04-16 20:50:31   
23356  {"Start Date":"2022-04-16","Renewal Date":"202... 2022-04-16 20:50:39   

      event_platform device_manufacture device_model  month  
0            android            samsung        SM-N9      1  
1            android            samsung        SM-N9      1  
2            android            samsung        SM-G9      1  
3            android            samsung        SM-N9      1  
4                ios              Apple     iPhone12      1  
...              ...                ...          ...    ...  
23352        android             Google      Pixel 3      4  
23353        android             Google      Pixel 3      4  
23354        android             Google      Pixel 3      4  
23355        android             Google      Pixel 3      4  
23356      outofband          outofband    outofband      4  

[22039 rows x 9 columns]

In [23]:
# export data
data.to_csv('processed_data.csv', encoding='utf-8', index=False)

### Interim findings
We processed data deleting duplicate values, filling missing values, adding some parameters. We have familiarized ourselves with the dataset and are now ready to work with it in the future

[See notebook Vizualization_plus_attributes]